src/Hedgehog/Internal/Function.idr

```diff
@@ -0,0 +1,351 @@
+module Hedgehog.Internal.Function
+
+import Data.Colist
+import Data.Cotree
+import Data.Either
+import Data.String
+
+import Hedgehog.Internal.Gen
+import Hedgehog.Internal.Range
+import Hedgehog.Internal.Seed
+
+import Language.Reflection.Util
+
+%default total
+
+public export
+interface Cogen a where
+  constructor MkCogen
+  perturb : a -> Seed -> Seed
+
+||| This function is meant to be used between successive perturbations is different arguments of the same constructor
+export
+shiftArg : Seed -> Seed
+shiftArg = variant 33 . snd . split . variant 31
+
+export
+cogen : Cogen a => a -> Gen b -> Gen b
+cogen x g = MkGen $ \sz, sd => unGen g sz $ perturb x sd
+
+export
+Cogen Unit where
+  perturb _ = id
+
+export
+Cogen (Equal x y) where
+  perturb Refl = id
+
+export
+Cogen Bool where
+  perturb True  = variant 0
+  perturb False = variant 1
+
+export
+Cogen Nat where
+  perturb = variant . cast
+
+export
+Cogen Char where
+  perturb = variant . cast
+
+export
+Cogen Void where
+  perturb _ impossible
+
+export
+Cogen a => Cogen b => Cogen (a, b) where
+  perturb (x, y) = perturb x . shiftArg . perturb y . shiftArg
+
+export
+Cogen a => Cogen b => Cogen (Either a b) where
+  perturb $ Left  x = perturb x . shiftArg . variant 0
+  perturb $ Right y = perturb y . shiftArg . variant 1
+
+export
+Cogen a => Cogen (Maybe a) where
+  perturb Nothing  = variant 0
+  perturb (Just x) = perturb x . shiftArg . variant 1
+
+export
+Cogen a => Cogen (List a) where
+  perturb []      = variant 0
+  perturb (x::xs) = perturb xs . shiftArg . perturb x . shiftArg . variant 1
+
+Integral a => Neg a => Ord a => Cast a (Bool, List Bool) where
+  cast n = if n >= 0 then (True, go [] n) else (False, go [] $ -n - 1) where
+    go : List Bool -> a -> List Bool
+    go bits x = if x == 0 then bits else go ((mod x 2 == 1) :: bits) (assert_smaller x $ div x 2)
+
+Integral a => Neg a => Cast (Bool, List Bool) a where
+  cast (sign, bits) = do
+    let body = foldl (\acc, b => acc * the a 2 + if b then 1 else 0) (the a 0) bits
+    if sign then body else negate $ body + 1
+
+export
```

Comment:
Could you please elaborate, why we use this implementation for all integral types with the exception of `Nat`? Would it not make more sense to injectively convert `Integer` to `Nat` and use `Cogen Nat` instead (see below)? All other integral types such as `BitsXY` and `IntXY` could go via `Integer`.

A possible injection (no loss of information) would map positive integers to even natural numbers and negative integers to odd ones: `if n >= 0 then cast (2 * n) else cast (2 * negate n + 1)`.

Reply:
> Could you please elaborate, why we use this implementation for all integral types with the exception of `Nat`?

The reasons are that 1) `Nat` does not implement `Integral` 2) `String` does implement `Cast String Nat` in a way that disallows be to use `Cast a Nat => Cogen a` 3) I didn't know about your interface `ToInteger` 4) These instances are anyway needed for `ShrCogen`, thus I just reused them.

Now I see that `ToInteger` is, probably, the best interface for this, except for supporting `Double`s in non-optimal way. `Integral` in this case is again slightly better. To deal with it, we can require both `ToInteger` and `Integral`, WDYT?

Reply:
Personally, I'd prefer to not use a generic numeric implementation at all but define explicit implementations for all primitive types. My experience with generic implementations is - for the reasons you just listed - not a good one, unless the generic implementation truly holds for all cases. So, I'd do the following

```idris
Cogen Bits64 where
  perturb = variant

Cogen Bits64 where
  perturb n =
    if n >= 0
     then perturb (cast {to = Bits64} $ 2*n)
     else perturb (cast {to = Bits64} $ 2 * negate n + 1)

Cogen Nat where
  perturb = perturb . cast {to = Bits64}

Cogen Bits8 where
  perturb = perturb . cast {to = Bits64}

Cogen Int8 where
  perturn = perturb . cast {to = Integer}
```

For me, this is the last thing blocking this PR. After this, I am ready to merge.

Reply:
I did it this way. But during this change I thought that `Bits` interface can be the best of both worlds, at least `FiniteBits`.

Reply:
Oh, and if you wonder why `ShrCogen Nat` was done through `ShrCogen Integer`, and not separately, the reasons are 1) not to repeat similar code 2) I checked that current approach is must faster than special through `divmodNatNZ`